### ISSUE TITLE
Question about MRP loss implementation

### ISSUE THREAD

**Issue by ahnjaewoo** [thunlp/CorefBERT#7]
Hi,
Thanks for such a great paper and open-sourced codes!
I have some questions about MRP loss implementation in your codes.

1. I cannot find equation (3), which refers to the multiplication of start & end tokens in CorefBERT paper, in your codes! Instead, I can only see start token embedding here (https://github.com/thunlp/CorefBERT/blob/f7db46ee1529ba1caede1f46c2ecc5e4c76f62c0/transformers/modeling_bert.py#L1365)
2. Why do you use `scores` variable to get `log_correct` variable? It seems weird to get gold truth values using predicted values inside. Am I missing something? https://github.com/thunlp/CorefBERT/blob/f7db46ee1529ba1caede1f46c2ecc5e4c76f62c0/transformers/modeling_bert.py#L1373

Thanks in advance!

**Comment by YeDeming** [comment]
Thanks for you interest.

1. We adopt the start and end tokens in the preprocessing code. And then we name both of them "span_start". 

https://github.com/thunlp/CorefBERT/blob/f7db46ee1529ba1caede1f46c2ecc5e4c76f62c0/Pretrain/gen_copy_data_MRP.py#L142-L145

Sorry for the confusion.

2. There are multiple gold truth positions since a word have mutiple mentions. We set the gold positions in label_mask to 1 to aquire log_correct  = log sum(exp(gold position 1) + exp(gold position 2) + ...).

Best,
Deming

**Comment by ahnjaewoo** [comment]
Thanks for your quick & kind response!
1. Okay I got it. To sum it up, all of the start & end tokens of referring words are jointly trained via MRP loss.
2. Oh sorry, but my question was why did you put `scores` variable to get gold truth values? Why not this?
`log_correct = torch.logsumexp(-10000 * (1 - label_masks), dim=-1)`

Thanks!

**Comment by YeDeming** [comment]
Sorry I don't get your question, E.g., when label_masks = [0 1 0],   log_correct  = torch.logsumexp(-10000 * (1 - label_masks), dim=-1) = log (exp(-10000)+exp(0)+exp(-10000)) = 0. And it doen't depend on the network output.

**Comment by ahnjaewoo** [comment]
Sorry, I will clarify my question again.

Gold truth label vector in your code is as below:
`log_correct = torch.logsumexp(scores - 10000 * (1 - label_masks), dim=-1) `

and this value depends on the predicted value (i.e. network output `scores`) which means the same label_masks [0,1,0] causes different loss value every time, because `scores` variable causes different loss value every time when we compute the forward loop.

This seems unnatural because the label needs to be independent of predicted values.
For instance, we calculate MLM (masked language modeling) loss not with a dynamic label vector (dependent of network output) but with a static one-hot vector.

**Comment by YeDeming** [comment]
Hi,

https://github.com/thunlp/CorefBERT/blob/f7db46ee1529ba1caede1f46c2ecc5e4c76f62c0/transformers/modeling_bert.py#L1372-L1375

`loss = log_correct - log_norm`

log_correct denotes the log of  the numerator of Equation (2). log_norm denotes the log of the denominator of Equation (2).

Best,
Deming

**Comment by ahnjaewoo** [comment]
Hi,
Thank you again for the clarification!

I was totally confused (because of the name `correct`), and thought that `log_norm` refers to the log of Equation (2) and `log_correct` refers to the log of the ground-truth label (which is wrong because the `copy_loss` term does not use cross-entropy loss!)

To sum it up, thank you and sorry for asking such an easy question.
Your comments helped me a lot!
